### PR TITLE
Add sortable columns to spreadsheet view, closes #291

### LINKED ISSUE #291: Spreadsheet - sort corps by Market value
To allow for quickly identifying corp OR turn order. Or possibly offer different sort option by clicking column headers.

### PR DESCRIPTION
**Pull request by wheresvic**
This PR allows sorting columns in the spreadsheet view. At the moment the following 3 columns are supported:
- Operating Order
- Cash
- Market price

The default is to sort by company symbol in ascending order (current view). The sort preferences are locally stored so they should be maintained over sessions :).

See screencast: [https://streamable.com/oobcg6](https://streamable.com/oobcg6)

### CODE REVIEW

File: assets/app/view/spreadsheet.rb [hunk 1]
```diff
@@ -79,14 +83,58 @@ def render_title
           h(:th, { attrs: { colspan: or_history_titles.size } }, 'OR History'),
           ]),
         h(:tr, [
-          h(:th, { style: { width: '20px' } }, ''),
+          h(:th, { style: { width: '20px' } }, [
```
Comment (tobymao) [inline assets/app/view/spreadsheet.rb:86]: you should be able to dry this up

`def sort_link(...)`

also lambda should be multiline do end, you don't have to inline it in the new function you create when you dry this up

File: assets/app/view/spreadsheet.rb [hunk 2]
```diff
@@ -2,10 +2,14 @@
 
 require 'view/stock_market'
 require 'lib/color'
+require 'view/link'
+require 'lib/storage'
 
 module View
   class Spreadsheet < Snabberb::Component
     needs :game
+    needs :spreadsheet_sort_by, default: Lib::Storage['spreadsheet_sort_by'] || 'ID', store: true
```
Comment (tobymao) [inline assets/app/view/spreadsheet.rb:11]: i made this framework and didn't even know you could do this lol.. probably safer to default to nil and then to reset it in the render

Comment (tobymao) [inline assets/app/view/spreadsheet.rb:11]: you're already using local storage, so there's no reason to store it here,

change these two needs to

needs sort_by default: nil
needs sort_order default: nil

File: assets/app/view/spreadsheet.rb [hunk 3]
```diff
@@ -96,14 +144,62 @@ def render_title
...
+    def sort_order_icon
+      return '(⬇️)' if @spreadsheet_sort_order == 'ASC'
+
+      '(⬆️)'
+    end
+
+    def mark_sort_column(sort_by)
+      Lib::Storage['spreadsheet_sort_by'] = sort_by
+      store(:spreadsheet_sort_by, sort_by)
+      *@spreadsheet_sort_by = sort_by
```
Comment (tobymao) [inline assets/app/view/spreadsheet.rb:156]: store will set the variable and rerender, you should not mutate the variable

File: assets/app/view/spreadsheet.rb [hunk 4]
```diff
@@ -96,14 +144,62 @@ def render_title
...
     end
 
+    def sort_order_icon
+      return '(⬇️)' if @spreadsheet_sort_order == 'ASC'
+
+      '(⬆️)'
+    end
+
+    def mark_sort_column(sort_by)
+      Lib::Storage['spreadsheet_sort_by'] = sort_by
```
Comment (tobymao) [inline assets/app/view/spreadsheet.rb:154]: i don't think you need any needs if you're using local stoarge

so just do

Lib::Storage['spreadsheet_sort_by'] = sort_by
update

do you want this to be user global or per game? i think user global makes ssense

File: assets/app/view/spreadsheet.rb [hunk 5]
```diff
@@ -96,14 +144,62 @@ def render_title
...
       floated_corporations = @game.round.entities
+
       @game.corporations.map do |c|
         operating_order = (floated_corporations.find_index(c) || -1) + 1
-        render_corporation(c, operating_order, current_round)
+        result << [operating_order, c]
+      end
+
+      result = result.sort_by do |c|
+        if @spreadsheet_sort_by == 'OPERATING-ORDER'
```
Comment (tobymao) [inline assets/app/view/spreadsheet.rb:190]: case @sort_by
when 'OPERATING-ORDER'
...

File: assets/app/view/spreadsheet.rb [hunk 6]
```diff
@@ -96,14 +144,62 @@ def render_title
...
+        elsif @spreadsheet_sort_by == 'CASH'
+          c[1].cash
+        elsif @spreadsheet_sort_by == 'SHARE-PRICE'
+          c[1].share_price.nil? ? 0 : c[1].share_price.price
+        else
+          c[1].id
+        end
       end
+
+      result = result.reverse if @spreadsheet_sort_order == 'DESC'
```
Comment (tobymao) [inline assets/app/view/spreadsheet.rb:201]: result.reverse! if ...

File: assets/app/view/spreadsheet.rb (hunk 2 above)
Comment (wheresvic) [reply assets/app/view/spreadsheet.rb:11]: yes, so it was the `update` function that I was missing! Will get rid of the `store`

Comment (wheresvic) [reply assets/app/view/spreadsheet.rb:11]: agreed.

File: assets/app/view/spreadsheet.rb (hunk 4 above)
Comment (wheresvic) [reply assets/app/view/spreadsheet.rb:154]: I think global also makes sense. So yeah `update` was the trick which I was missing - will clean it up!

File: assets/app/view/spreadsheet.rb (hunk 3 above)
Comment (wheresvic) [reply assets/app/view/spreadsheet.rb:156]: yep!

File: assets/app/view/spreadsheet.rb (hunk 6 above)
Comment (wheresvic) [reply assets/app/view/spreadsheet.rb:201]: cool

File: assets/app/view/spreadsheet.rb (hunk 1 above)
Comment (wheresvic) [reply assets/app/view/spreadsheet.rb]: done!

File: assets/app/view/spreadsheet.rb (hunk 5 above)
Comment (wheresvic) [reply assets/app/view/spreadsheet.rb]: done

File: assets/app/view/spreadsheet.rb [hunk 7]
```diff
@@ -2,12 +2,19 @@
 
 require 'view/stock_market'
 require 'lib/color'
+require 'view/link'
+require 'lib/storage'
 
 module View
   class Spreadsheet < Snabberb::Component
     needs :game
+    needs :spreadsheet_sort_by, default: nil
```
Comment (tobymao) [inline assets/app/view/spreadsheet.rb:11]: not needed, delete

File: assets/app/view/spreadsheet.rb [hunk 8]
```diff
@@ -79,14 +86,14 @@ def render_title
...
-          h(:th, { style: { width: '20px' } }, ''),
+          render_sort_link({ style: { width: '20px' } }, 'SYM', 'ID'),
           *@game.players.map { |p| h(:th, props, p.name) },
           h(:th, props, 'IPO'),
           h(:th, props, 'Market'),
           h(:th, props, 'IPO'),
-          h(:th, props, 'Market'),
-          h(:th, props, 'Cash'),
-          h(:th, props, 'Operating Order'),
+          render_sort_link(props, 'Market', 'SHARE-PRICE'),
```
Comment (tobymao) [inline assets/app/view/spreadsheet.rb:94]: let's call it SHARE_PRICE and OPERATING_ORDER

File: assets/app/view/spreadsheet.rb [hunk 9]
```diff
@@ -96,14 +103,71 @@ def render_title
...
+      Lib::Storage['spreadsheet_sort_order'] = 'ASC' if @spreadsheet_sort_order == 'DESC'
+      Lib::Storage['spreadsheet_sort_order'] = 'DESC' unless @spreadsheet_sort_order == 'DESC'
+      update
+    end
+
     def render_corporations
       current_round = @game.round.turn_round_num
 
+      ordered_corporations = sorted_corporations
+      ordered_corporations.map do |c|
```
Comment (tobymao) [inline assets/app/view/spreadsheet.rb:142]: sorted_corporations.map

don't need the extra var ordered_corporations

File: assets/app/view/spreadsheet.rb [hunk 10]
```diff
@@ -96,14 +103,71 @@ def render_title
...
+        render_corporation(c[1], c[0], current_round)
+      end
+    end
+
+    def sorted_corporations
+      result = []
       floated_corporations = @game.round.entities
+
       @game.corporations.map do |c|
```
Comment (tobymao) [inline assets/app/view/spreadsheet.rb:151]: if you're doing map, there's no reason to have an array

result = @game.corporations.map do |c|
[operating_order, c]
end

File: assets/app/view/spreadsheet.rb [hunk 11]
```diff
@@ -96,14 +103,71 @@ def render_title
...
         operating_order = (floated_corporations.find_index(c) || -1) + 1
-        render_corporation(c, operating_order, current_round)
+        result << [operating_order, c]
       end
+
+      result = result.sort_by do |c|
```
Comment (tobymao) [inline assets/app/view/spreadsheet.rb:156]: you can just do result.sort_by! without assignng a new var

Comment (tobymao) [inline assets/app/view/spreadsheet.rb:156]: result.sort_by! do |order, corporation|

File: assets/app/view/spreadsheet.rb [hunk 12]
```diff
@@ -96,14 +103,71 @@ def render_title
...
+        case @spreadsheet_sort_by
+        when 'OPERATING-ORDER'
+          c[0]
+        when 'CASH'
+          c[1].cash
+        when 'SHARE-PRICE'
+          c[1].share_price.nil? ? 0 : c[1].share_price.price
```
Comment (tobymao) [inline assets/app/view/spreadsheet.rb:163]: corporation.share_price&.price || 0

File: assets/app/view/spreadsheet.rb (hunk 7 above)
Comment (wheresvic) [reply assets/app/view/spreadsheet.rb:11]: whoops - will remove!

File: assets/app/view/spreadsheet.rb (hunk 8 above)
Comment (wheresvic) [reply assets/app/view/spreadsheet.rb:94]: sure

File: assets/app/view/spreadsheet.rb (hunk 12 above)
Comment (wheresvic) [reply assets/app/view/spreadsheet.rb:163]: got it

**Pull request merged**